src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/SmartContractPosBlockSignatureRule.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using NBitcoin.Crypto;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.Consensus;
+using Stratis.Bitcoin.Features.Consensus.Rules;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Consensus.Rules
+{
+    /// <summary>
+    /// A rule that will validate the signature of a PoS block.
+    /// </summary>
+    [PartialValidationRule(CanSkipValidation = true)]
+    [IntegrityValidationRule]
+    public class SmartContractPosBlockSignatureRule : StakeStoreConsensusRule
+    {
+        /// <inheritdoc />
+        /// <exception cref="ConsensusErrors.BadBlockSignature">The block signature is invalid.</exception>
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.ValidationContext.Block;
+
+            if (!(block is SmartContractPosBlock posBlock))
```

Comment:
if you are gonna throw invalid cast exception, why not just try to cast it?

Reply:
@majikandy please note that I haven't gone into a refactoring exercise at all for this work and this will only be merged to our sc-dev branch, not master. The code in the PR will have some duplication so we will look at bettering this at another time. (Copy paste code from the FN)

Reply:
Ah ok, in that case i’ll stop :)